File: survival-common/src/main/java/net/survival/block/Column.java
package net.survival.block;

import java.util.Stack;

public class Column {
    public static final int XLENGTH = Chunk.XLENGTH;
    public static final int ZLENGTH = Chunk.ZLENGTH;
    public static final int BASE_AREA = XLENGTH * ZLENGTH;

    private final Stack<Chunk> chunks;
    private boolean newFlag = true;
    private boolean modified;

    public Column() {
        chunks = new Stack<>();
    }

    private Column(Stack<Chunk> chunks) {
        this.chunks = chunks;
    }

    public Column makeCopy() {
        var copyOfChunks = new Stack<Chunk>();
        for (var chunk : chunks) {
            copyOfChunks.push(chunk.makeCopy());
        }

        return new Column(copyOfChunks);
    }

    public Chunk getChunk(int index) {
        if (index >= chunks.size()) {
            return null;
        }

        return chunks.get(index);
    }

    public Chunk getChunkLazy(int index) {
        while (index >= chunks.size()) {
            chunks.push(new Chunk());
        }

        return chunks.get(index);
    }

    public Chunk getTopChunk() {
        return chunks.peek();
    }

    public void pushChunk(Chunk chunk) {
        chunks.push(chunk);
    }

    public int getHeight() {
        return chunks.size();
    }

    public Block getBlock(int x, int y, int z) {
        var index = y / Chunk.YLENGTH;

        if (index >= chunks.size()) {
            return StandardBlocks.AIR;
        }

        return chunks.get(index).getBlock(x, y % Chunk.YLENGTH, z);
    }

    public void setBlock(int x, int y, int z, Block to) {
        var index = y / Chunk.YLENGTH;

        while (index >= chunks.size()) {
            chunks.push(new Chunk());
        }

        chunks.get(index).setBlock(x, y % Chunk.YLENGTH, z, to);
        modified = true;
    }

    public boolean isInBounds(int lx, int ly, int lz) {
        return lx >= 0 && ly >= 0 && lz >= 0 && lx < XLENGTH && lz < ZLENGTH;
    }

    public boolean isNew() {
        return newFlag;
    }

    public void clearNewFlag() {
        newFlag = false;
    }

    public boolean isModified() {
        return modified;
    }

    public void clearModifiedFlag() {
        modified = false;
    }
}

File: survival-common/src/main/java/net/survival/block/Column.kt
package net.survival.block

import java.util.*

class Column {
    companion object {
        const val XLENGTH = Chunk.XLENGTH
        const val ZLENGTH = Chunk.ZLENGTH
        const val BASE_AREA = XLENGTH * ZLENGTH
    }

    private val chunks: ArrayList<Chunk>

    constructor() {
        chunks = ArrayList(8)
    }

    constructor(column: Column) {
        chunks = ArrayList(column.chunks.size)
        for (chunk in column.chunks) {
            chunks.add(Chunk(chunk))
        }
    }

    fun getChunk(index: Int): Chunk = chunks[index]
    fun getChunkOrNull(index: Int): Chunk? = chunks.getOrNull(index)
    fun addChunk(chunk: Chunk) {
        chunks.add(chunk)
    }

    val height: Int get() = chunks.size

    fun getBlock(x: Int, y: Int, z: Int): Block {
        val index = y / Chunk.YLENGTH
        return when {
            index >= chunks.size -> StandardBlocks.AIR
            else -> chunks[index].getBlock(x, y % Chunk.YLENGTH, z)
        }
    }

    fun setBlock(x: Int, y: Int, z: Int, block: Block) {
        val index = y / Chunk.YLENGTH
        while (index >= chunks.size) {
            chunks.add(Chunk())
        }
        chunks[index].setBlock(x, y % Chunk.YLENGTH, z, block)
    }

    fun isInBounds(lx: Int, ly: Int, lz: Int): Boolean =
        lx >= 0 && ly >= 0 && lz >= 0 && lx < XLENGTH && lz < ZLENGTH
}